refactor(api): axis and mounts for ot3

sfoster1 opened the pull request:
The OT3 has a different hardware configuration than the OT2. It has a different number of axes that have different names, and a different number of mounts. It's pretty annoying (and insufficient) to map the current axis and mount types into the OT3, and adding the OT3 stuff to the already existing axis and mounts pollutes the global API even when not using the OT3, confusing users.

This is a different approach. It creates OT3Mount and OT3Axis types, inside hardware_control.types. It characteristically makes input values to OT3API methods take unions of the old and new - e.g. `Union[Mount, OT3Mount]`, which is safe because it includes the API mandated by the hardware control protocols, but also allows a window to new functionality. Internally, the OT3API and backends use these new mount and axis types everywhere.

This has some followon requirements that are almost, but not quite, pure type-level changes. The instrument handler and the motion utilities are now generic across axis and mount types; this is usually sufficient but sometimes they need explicit enum type value callouts, which have to be provided or derived (at runtime level, not type level) from some other inputs. It's kind of messy and kind of a lot of code, but it does remove all the string values from the API/backend interface and lets us callout the new axis names from the top down.

Open to feedback about better ways this can be done.

sfoster1 committed:
refactor(api): axis and mount types for ot3

The OT3 has a different set of axes and mounts, and it's good not to
confuse them with the OT2. While for now the upwards interface is the
same, by using these types in the api<->controller interface we can
eliminate string roundtrips. They percolate down to the hardware package
for use as keys in the motion math position dictionaries.

There's also the introduction of the new concept of "Axis kind". This is
sort of a generalization of the old Axis.gantry_axes methods. Especially
as we start having grippers and other tools, we have a lot of different
things that fall under the category of "plunger axis" or "z axis", and
it's nice to be able to refer to them generically. One such place where
it's nice to refer to them generically is in axis constraints and config
parsing, where we really just want a setting that applies to _any_ Z
axis.

sfoster1 committed:
refactor(api): Use ot3axis and mount internally

So it's nice to be able to use the new types to talk to the backends but
if you have to switch them from the old ones as inputs, and to the old
ones as outputs, you're still really just washing data around for no
gain. There's not much we can do about the outputs unless OT3Axis
contravaries with Axis and OT3Mount contravaries with Mount - but,
Union[OT3Axis, Axis] covaries with Axis, and Union[OT3Mount, Mount]
covaries with mount, so we can add the option for callers to use the
new, more-capable enums when they interact with an ot3 hardware control
API.

This leaves us in an interesting conundrum, though. We want to use them
internally; we want to have them as inputs; and we want to share
functionality.

Luckily, the enums are plenty generic to be used as generics, since
they're only being used as keys, especially if you can retrieve the type
objects in the rare instances where necessary from preconfiguration. So
InstrumentHandler, the major shared-functionality point, gets to be a
generic, and its little worker dataclasses get to be generics, and
because you don't have type maps or anything you get a bunch of
overloads to indicate that Axis-OT3Axis and Mount-OT3Mount are linked,
and there's a whole bunch of code with the result that things nicely
vary.

The second problem though is that now, if you have InstrumentHandler
nice and generic, then OT3API can't inherit from it but switch out all
its methods to union types because that doesn't work. So now it's a
has-a relationship. Wheeeeee

amitlissack commented on api/src/opentrons/hardware_control/backends/utils.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def axis_to_node(axis: OT3Axis) -> "NodeId":
+    anm = {
```
anm?

sfoster1 replied:
axis node map, sorry

amitlissack commented on api/src/opentrons/hardware_control/motion_utilities.py:
```diff
@@ -1,79 +1,242 @@
+@lru_cache(4)
 def offset_for_mount(
-    primary_mount: Mount, left_mount_offset: Point, right_mount_offset: Point
+    primary_mount: Union[Mount, OT3Mount],
```
Sorry, but I am a bit baffled by this function. 

Is it the equivalent of `return left_mount_offset if primary_mount == LEFT else right_mount_offset`? 

Why the cache?

sfoster1 replied:
yeah it's the same, it just has to happen in a couple places and deal with several possible values for e.g. LEFT. Felt it was worth just having as a little function. The cache is some probably premature optimization

amitlissack commented on api/src/opentrons/hardware_control/types.py:
```diff
@@ -71,6 +71,116 @@ def __str__(self) -> str:
+        return top_types.Mount[self.name]
+
+
+class OT3Axis(enum.Enum):
```
I'd really prefer full names. But If you want to remain super brief can we slightly extend the names `Grip` vs `G`  or `Pip` vs `P`?

sfoster1 replied:
I suppose... I really like keeping axis names as math notation. It's the frame of mind.

amitlissack commented on api/src/opentrons/hardware_control/api.py:
```diff
@@ -59,7 +59,7 @@
     ExecutionManagerProvider,
     RobotCalibrationProvider,
-    InstrumentHandlerProvider,
+    InstrumentHandlerProvider[top_types.Mount],
```
I don't see that `InstrumentHandlerProvider` changed to generic. What am I missing?

sfoster1 commented on api/src/opentrons/hardware_control/instrument_handler.py:
```diff
@@ -34,23 +39,70 @@
+    ending_current: Dict[AxisType, float]
+
+
+class InstrumentHandlerProvider(Generic[MountType]):
```
@amitlissack here's where IHP became generic

amitlissack replied:
Oops. This file was collapsed.

amitlissack reviewed (approved):
I can't think of a better way to do this. Rock on!

ahiuchingau commented on api/src/opentrons/hardware_control/types.py:
```diff
@@ -71,6 +71,116 @@ def __str__(self) -> str:
+        return cls.Z_L, cls.Z_R
+
+    @classmethod
+    def mount_axes(cls) -> Tuple["OT3Axis", "OT3Axis"]:
```
`mount_axes` here returns the same axes as `pipette_axes`, is that intentional?

sfoster1 replied:
Yyyyesss. This is a thing I was going back and forth on. It may now be easier to do, but when first implementing this I noticed that you get a lot of gripper entries that crash things.

ahiuchingau reviewed (approved):
Got a question about pipette & mount axes. Everything else lgtm

Laura-Danielle commented on api/src/opentrons/hardware_control/api.py:
```diff
@@ -116,7 +116,9 @@ def __init__(
         ExecutionManagerProvider.__init__(self, loop, isinstance(backend, Simulator))
         RobotCalibrationProvider.__init__(self)
-        InstrumentHandlerProvider.__init__(self)
+        InstrumentHandlerProvider.__init__(
```
why do we need to pass this into the constructor?

sfoster1 replied:
because I hate super()

Laura-Danielle replied:
finneee

Laura-Danielle commented on api/src/opentrons/hardware_control/instrument_handler.py:
```diff
@@ -34,23 +39,70 @@
-        }
+    def __init__(self, attached_instruments: InstrumentsByMount[MountType]):
+        assert attached_instruments
+        self._attached_instruments: InstrumentsByMount[MountType] = attached_instruments
```
Why can't IHP build the mount dict itself?

sfoster1 replied:
Because then it would have to have runtime knowledge of which mount enum it's using. By passing this in from the caller, which does have that knowledge, we can avoid isinstance() checks; and by getting values by accessing the type of the keys of the dict inside IHP, we can avoid switching. It's kind of gross.

Laura-Danielle reviewed (approved):
Had a general question about the IHP class, but otherwise it looks good.

sfoster1 committed:
refactor(hardware): Parametrize motion axis types

This code previously used string axis names, which was fine, but it's
annoying to make it work with higher level code especially once that
higher level code transitions to axis enums.

The thing is, there's not really anything that this code _needs_ out of
any given axis name type. In 99% of cases it was just used as a dict
key, and the other cases were because of this Coordinates dataclass.

By making the Coordinates dataclass a generic type alias for a Mapping,
there's very few code changes needed (outside of tests, of course) to
make this code completely adaptable to basically any hashable type that
you feel like using as a key. Tests can use strings because it's easy;
and the api subpackage can use Axis, or OT3Axis, and all the typechecks
make it work.

amitlissack commented on api/src/opentrons/config/defaults_ot3.py:
```diff
@@ -27,197 +27,215 @@
 def _build_default_bpk(
-    from_conf: Any, default: ByGantryLoad[GeneralizeableAxisDict]
-) -> ByGantryLoad[GeneralizeableAxisDict]:
+    from_conf: Any, default: ByGantryLoad[Dict[OT3AxisKind, float]]
```
Can't you use `PerPipetteAxisSettings` instead of `ByGantryLoad[Dict[OT3AxisKind, float]]`?

amitlissack commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -49,7 +50,7 @@
 
 
-AxisValueMap = Dict[str, float]
+AxisValueMap = Dict[OT3Axis, float]
```
Seems we could be using this type alias in lots of places above.

sfoster1 replied:
It starts feeling a little pointless honestly especially if you make it a generic alias. Pushed a commit that does it.

sfoster1 committed:
Use some extra type aliases
